src/hotspot/share/runtime/arguments.cpp

```diff
@@ -3754,6 +3754,17 @@ jint Arguments::apply_ergo() {
   }
 #endif // COMPILER2_OR_JVMCI
 
+  if (log_is_enabled(Info, init)) {
+     FLAG_SET_ERGO_IF_DEFAULT(ProfileClassLinkage, true);
+  }
```

Comment:
What if ProfileClassLinkage is set true on the command-line without -Xlog:init? That doesn't seem to make sense to me. So I'm not clear why it is a settable diagnostic flag.

Reply:
If only `ProfileClassLinkage` is set to true without `-Xlog:init`, the user will not see any counters output.
In `java.cpp`:
```
160 void log_vm_init_stats() {
161   LogStreamHandle(Info, init) log;
162   if (log.is_enabled()) {
163     ClassLoader::print_counters();
164   }
165 }
```

In the future, there will be other sets of counters controlled by other diagnostic flags.

Reply:
We are planning to add more diagnostic flags to control different sets of counters. With the current design, the user just needs to specify `-Xlog:init` to enable all the "new" counters. If the `init` logging is enabled by individual flag, the user needs to enable individual flag in the command line.
Anyway, I think the follow would achieve what you are alluding to?
```
  if (FLAG_IS_CMDLINE(ProfileClassLinkage) && !log_is_enabled(Info, init)) {
    LogConfiguration::configure_stdout(LogLevel::Info, true, LOG_TAGS(init));
  }
```
I think it's better to keep the current change.

Reply:
This still seems convoluted to me. A -Xlog option shouldn't control anything but logging. If you want a set of counters enabled then use the flag to enable them, and separately use -Xlog:init to print them (though whether "init" is appropriate here is another matter). You could use -Xlog:init+foo to be more selective about which counters.

Reply:
I've modified the fix so that the user needs to specify both `-Xlog:init` and `-XX:+ProfileClassLinkage` for the counters to be printed.

Reply:
Thanks @iwanowww for charming in.
I've changed `arguments.cpp` back to the first version.

Reply:
I'm getting confused.

The ProfileClassLinkage flag should control whether the counters are initialized and used.

The -Xlog:init (perhaps with a better name/tag!) should control whether they get printed - in theory you could chose to print under any logging setting.

I don't think either flag should imply/force the setting of the other.

Reply:
It's about usability. `-Xlog:init` means "I want to see logs related to initialization", so it should enable all the counters for printing the related logs. We may add several groups of counters in the future. We don't want to force the user to enumerate all these counters like:

```
java -Xlog:init -XX:+ ProfileClassLinkage -XX:+ProfileAAA  -XX:+ProfileBBB ....
```

Also,  `-Xlog:init` doesn't force the all the counters to be used. It just changes the default value of the selection:

```
  if (log_is_enabled(Info, init)) {
    FLAG_SET_ERGO_IF_DEFAULT(ProfileClassLinkage, true);
    FLAG_SET_ERGO_IF_DEFAULT(ProfileAAA, true);
    FLAG_SET_ERGO_IF_DEFAULT(ProfileBBB, true);
  }
```

The user could disable certain counters by:

```
java -Xlog:init  -XX:-ProfileAAA
```

Reply:
> The -Xlog:init (perhaps with a better name/tag!)

I'm all for a better naming scheme. Any suggestions?

Reply:
> -Xlog:init means "I want to see logs related to initialization", so it should enable all the counters for printing the related logs. 

I don't agree. Initialization logging could encompass many different things, some of which are individually controllable via different flags. Simply turning on init logging should not turn on all such flags. If you want that level of coupling then perhaps use init_counters (or something like that) to make it clear this is not a general log tag intended for any initialization code to use, but something you have chosen to tie to specific functionality.

> We may add several groups of counters in the future. We don't want to force the user to enumerate all these counters

It is not clear to me how you envisage that working. You want individual group switches plus a global one?

Reply:
OK, I agree that not `-Xlog:init` logs are related to timing. 

We actually have an existing tag that's used for conditional logging. Perhaps we should use that instead?

```
./share/logging/logTag.hpp:  LOG_TAG(startuptime) \
./share/memory/universe.cpp:  TraceTime timer("Genesis", TRACETIME_LOG(Info, startuptime));
./share/prims/methodHandles.cpp:  TraceTime timer("MethodHandles adapters generation", TRACETIME_LOG(Info, startuptime));
./share/runtime/stubRoutines.cpp:  TraceTime timer(timer_msg, TRACETIME_LOG(Info, startuptime));
./share/runtime/threads.cpp:  TraceTime timer("Initialize module system", TRACETIME_LOG(Info, startuptime));
./share/runtime/threads.cpp:  TraceTime timer("Initialize java.lang classes", TRACETIME_LOG(Info, startuptime));
./share/runtime/threads.cpp:  TraceTime timer("Initialize java.lang.invoke classes", TRACETIME_LOG(Info, startuptime));
./share/runtime/threads.cpp:  TraceTime timer("Create VM", TRACETIME_LOG(Info, startuptime));
./share/runtime/threads.cpp:  { TraceTime timer("Start VMThread", TRACETIME_LOG(Info, startuptime));
./share/runtime/timerTrace.hpp://    TraceTime t("some timer", TIMERTRACE_LOG(Info, startuptime, tagX...));
./share/utilities/ostream.cpp:  // lazily create log file (at startup, LogVMOutput is false even

/jdk3/zoo/open/src/hotspot$ tbjava -Xlog:startuptime --version
[0.010s][info][startuptime] StubRoutines generation initial stubs, 0.0006132 secs
[0.025s][info][startuptime] Genesis, 0.0145142 secs
[0.025s][info][startuptime] StubRoutines generation continuation stubs, 0.0000198 secs
[0.028s][info][startuptime] Interpreter generation, 0.0005919 secs
[0.028s][info][startuptime] StubRoutines generation final stubs, 0.0000717 secs
[0.028s][info][startuptime] MethodHandles adapters generation, 0.0000101 secs
[0.029s][info][startuptime] Start VMThread, 0.0000846 secs
[0.032s][info][startuptime] Initialize java.lang classes, 0.0036507 secs
[0.033s][info][startuptime] Initialize java.lang.invoke classes, 0.0002106 secs
[0.034s][info][startuptime] StubRoutines generation compiler stubs, 0.0015061 secs
[0.035s][info][startuptime] Initialize module system, 0.0017471 secs
[0.035s][info][startuptime] Create VM, 0.0305224 secs
java 23-internal 2024-09-17
Java(TM) SE Runtime Environment (build 23-internal-adhoc.iklam.zoo)
Java HotSpot(TM) 64-Bit Server VM (build 23-internal-adhoc.iklam.zoo, mixed mode, sharing)
```

Reply:
The `TraceTime` doesn't fit what we need well. All the function using `TraceTime `will be called only once.
I tried it in `LinkResolver::resolve_invokehandle`
```
void LinkResolver::resolve_invokehandle(CallInfo& result, const constantPoolHandle& pool, int index, TRAPS) {

  TraceTime trace_timer("Resolve invokehandle", TRACETIME_LOG(Info, startuptime));
```
When I run a HelloWorld with `-Xlog:startuptime`, I saw:
```
[0.008s][info][startuptime] StubRoutines generation initial stubs, 0.0005199 secs
[0.060s][info][startuptime] Genesis, 0.0519067 secs
[0.060s][info][startuptime] StubRoutines generation continuation stubs, 0.0000791 secs
[0.069s][info][startuptime] Interpreter generation, 0.0022987 secs
[0.082s][info][startuptime] StubRoutines generation final stubs, 0.0012770 secs
[0.082s][info][startuptime] MethodHandles adapters generation, 0.0001008 secs
[0.082s][info][startuptime] Start VMThread, 0.0001566 secs
[0.092s][info][startuptime] Initialize java.lang classes, 0.0101517 secs
[0.094s][info][startuptime] Initialize java.lang.invoke classes, 0.0007457 secs
[0.097s][info][startuptime] StubRoutines generation compiler stubs, 0.0035488 secs
[0.100s][info][startuptime] Initialize module system, 0.0064275 secs
[0.101s][info][startuptime] Create VM, 0.1007149 secs
[0.135s][info][startuptime] Resolve invokehandle, 0.0004090 secs
[0.135s][info][startuptime] Resolve invokehandle, 0.0000021 secs
[0.135s][info][startuptime] Resolve invokehandle, 0.0000009 secs
[0.137s][info][startuptime] Resolve invokehandle, 0.0002939 secs
[0.137s][info][startuptime] Resolve invokehandle, 0.0000017 secs
[0.139s][info][startuptime] Resolve invokehandle, 0.0001010 secs
[0.139s][info][startuptime] Resolve invokehandle, 0.0000016 secs
[0.143s][info][startuptime] Resolve invokehandle, 0.0000026 secs
hello, world
```
Also, I think `TraceTime` doesn't have a counter to track the number of times the function is called.

Reply:
> > The -Xlog:init (perhaps with a better name/tag!)
> 

How about `-Xlog:initcounters` as the name/tag? It is similar to @dholmes-ora 's suggestion above without the underscore between "init" and "counters". Currently, none of the logging tag contains an underscore character.

@iwanowww, @iklam, @dholmes-ora, what do you guys think?

Reply:
While these counters are useful for start-up measurement, they can be used for other purposes (e.g., monitoring how much time is spend in linking classes in a long running application). So I think the logging tag should be more neutral.

We already have `-Xlog:perf`. Maybe we can have sub options like `-Xlog:perf+class+link` for the counters in this PR?

I know an upcoming PR will be for stats for MutexLocker. Something like `-Xlog:perf+lock` would work for such stats.

Reply:
The above suggestion is fine with me. The `perf` tag is pre-existence, I noticed that if I specified `-Xlog:perf*`, the following log output appeared:
```
[0.002s][info][perf,memops] Trying to open /tmp/...
[0.003s][info][perf,memops] Successfully opened
```
It is probably ok if there are not too many of the above log output.

I've made the following changes and will push another commit after some testing:
- added the `link` logging tag; both `perf` and `class` tags already there;
- removed the `ProfileClassLinkage` diagnostics flag;
- added a global bool variable `_perf_class_link` in arguments.hpp;
  it will be set to true if `-Xlog:perf+class+link` is specified. I think it is more efficient to check `Argument::perf_class_link()` vs `log_is_enabled(Info, perf, class, link)`
- changed the check for `log_is_enabled(Info, init)` to `log_is_enabled(Info, perf, class, link)`.

Reply:
@dholmes-ora, @iklam Could you review my latest commit?
Thanks!

Reply:
The latest version looks good to me. My only suggestion is to use `log_is_enabled(Info, perf, class, link)` directly, because it's very efficient.

To verify, I wrote a function like this:

```
void foo() {
  if (log_is_enabled(Info, class, load)) {
    tty->print_cr("Hello");
  }
}
```

GCC compiles it to the following. So the test is implemented as a single load from memory.

```
_Z3foov:
.LFB8725:
	.cfi_startproc
	movq	64+_ZN16LogTagSetMappingILN6LogTag4typeE16ELS1_74ELS1_0ELS1_0ELS1_0ELS1_0EE7_tagsetE(%rip), %rax
	testq	%rax, %rax
	je	.L6
	movq	tty@GOTPCREL(%rip), %rax
	leaq	.LC0(%rip), %rsi
	movq	(%rax), %rdi
	xorl	%eax, %eax
	jmp	_ZN12outputStream8print_crEPKcz@PLT
	ret
```

Reply:
> The latest version looks good to me. My only suggestion is to use `log_is_enabled(Info, perf, class, link)` directly, because it's very efficient.
> 

Thanks for checking it.
I also did some perf testing using `log_is_enabled(Info, perf, class, link)` and the results are similar to using `Arguments::perf_class_link()`. So I've pushed another commit based on your suggestion.